pkg/disk/group_volume_snapshot_utils.go

```diff
@@ -0,0 +1,259 @@
+func ifExistsGroupSnapshotMatchSourceVolume(existsGroupSnapshot *ecs.SnapshotGroup, sourceVolumeIds []string) bool {
+	if len(existsGroupSnapshot.Snapshots.Snapshot) != len(sourceVolumeIds) {
+		return false
+	}
+	existsSnapshotSourceIdsMap := make(map[string]struct{})
+	for _, snapshot := range existsGroupSnapshot.Snapshots.Snapshot {
+		existsSnapshotSourceIdsMap[snapshot.SourceDiskId] = struct{}{}
+	}
+	matach := true
+	for _, id := range sourceVolumeIds {
+		if _, ok := existsSnapshotSourceIdsMap[id]; !ok {
+			matach = false
+			break
+		}
+	}
+	return matach
+}
+
+// ifExistsGroupSnapshotMatch checks whether there snapshotIDs in `snapshotIds`
+// that correspond to all related snapshots in `existsGroupSnapshot`.
+// If `requireExactMatch` is true, the function requires an exact match between
+// the related snapshotIDs and `snapshotIds`, including the same number of elements.
+// If `requireExactMatch` is false, it only checks if each related snapshotID
+// can be found in the `snapshotIds`, that means actual snapshots in group can be a subset of `snapshotIds`
+func ifExistsGroupSnapshotMatch(existsGroupSnapshot *ecs.SnapshotGroup, snapshotIds []string, requireExactMatch bool) bool {
+	if requireExactMatch && len(snapshotIds) != len(existsGroupSnapshot.Snapshots.Snapshot) {
+		return false
+	}
+	sourceSnapshotIdMap := make(map[string]struct{})
+	for _, id := range snapshotIds {
+		sourceSnapshotIdMap[id] = struct{}{}
+	}
+	matach := true
+	for _, snapshot := range existsGroupSnapshot.Snapshots.Snapshot {
+		if _, ok := sourceSnapshotIdMap[snapshot.SnapshotId]; !ok {
+			matach = false
+			break
+		}
+	}
+	return matach
+}
+
+// checkSourceVolumes checks whether a single source volume is valid.
+func checkSourceVolume(disk *ecs.Disk, zoneId string, capacityInGiB int) (
+	nZoneId string, nCapacityInGiB int, err error) {
+	nZoneId = zoneId
+	nCapacityInGiB = capacityInGiB
+
+	switch disk.Category {
+	case string(DiskESSD), string(DiskESSDAuto):
+	default:
+		err = fmt.Errorf("groupSnapshot only support ESSD disks, but disk %s is %v", disk.DiskId, disk.Category)
+		return
+	}
+
+	if nZoneId == "" {
+		nZoneId = disk.ZoneId
+	}
+	if nZoneId != disk.ZoneId {
+		err = fmt.Errorf("groupSnapshot only support disks from one zone, but get disks from %s and %s", zoneId, disk.ZoneId)
+		return
+	}
+
+	if disk.MultiAttach != DiskMultiAttachDisabled {
+		err = fmt.Errorf("groupSnapshot do not support multiAttached disk, but disk %s enables multiAttached", disk.DiskId)
+		return
+	}
+
+	nCapacityInGiB += disk.Size
+	if nCapacityInGiB > MaxCapacityInGiB {
+		err = fmt.Errorf("total capacity is larger than %d", MaxCapacityInGiB)
+		return
+	}
+	return
+}
+
+// checkSourceVolumes checks whether the source volumes are valid.
+func checkSourceVolumes(sourceVolumeIds []string) error {
+	var zoneId string
+	var capacityInGiB int
+	var err error
+	for _, sourceVolumeId := range sourceVolumeIds {
+		disks := getDisk(sourceVolumeId, GlobalConfigVar.EcsClient)
```

Comment:
Retrieve all the disk in one DescribeDisks request? BTW, do we really need to do all these validations? Why not just let ECS do the validation and we just check the error code.